File: day01.py
"""Solution to day 1 for Advent of Code"""

from get_input import get_input


def part1(elves):
    return max(sum(elf) for elf in elves)


def part2(elves):
    return sum(sorted([sum(elf) for elf in elves])[-3:])


def parse(text):
    return [
        [int(n) for n in group.split("\n")]
        for group in text.strip().split("\n\n")
    ]


if __name__ == "__main__":
    LINES = parse(get_input(day=1, year=2022))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day01.rs
use common::get_input;
use std::collections::BinaryHeap;

type Input = Vec<Vec<usize>>;

fn part1(elves: &Input) -> usize {
    elves.iter().map(|elf| elf.iter().sum()).max().unwrap()
}

fn part2(elves: &Input) -> usize {
    elves.iter()
        .map(|elf| elf.iter().sum())
        .collect::<BinaryHeap<usize>>().iter()
        .take(3).sum()
}

fn parse(text: String) -> Input {
    text.trim().split("\n\n").map(|group| {
        group.split("\n").map(|n| n.parse::<usize>().unwrap()).collect()
    }).collect()
}

fn main() {
    let input = parse(get_input(01, 2022));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
